emerging_optimizers/psgd/psgd.py

```diff
@@ -0,0 +1,318 @@
+        updated_Q.append(q)
+        updated_L.append(l_updated)
+
+    return updated_Q, updated_L
+
+
+def _update_matrix_preconditioner(
+    q: torch.Tensor,
+    L: torch.Tensor,
+    precond_grad_cov: torch.Tensor,
+    total_numel: int,
+    precond_lr: float,
+    betaL: float,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Update matrix-structured preconditioner with adaptive Lipschitz constant.
+
+    Args:
+        q: Kronecker factor matrix for this dimension to update.
+        L: Lipschitz constant for this dimension.
+        precond_grad_cov: Gradient covariance.
+        total_numel: Total number of elements in the gradient.
+        precond_lr: Learning rate.
+        betaL: Beta for the Lipschitz constant exponential moving average.
+
+    Returns:
+        q: Updated Kronecker factor matrix for this dimension.
+        L: Updated Lipschitz constant for this dimension.
+    """
+    normalization = total_numel / q.shape[0]
+    ell = norm_lower_bound_spd(precond_grad_cov) + normalization
+    L = torch.max(betaL * L + (1 - betaL) * ell, ell)
+    q = q - precond_lr / L * (precond_grad_cov @ q - normalization * q)
+    q = procrustes_step(q)
+    return q, L
+
+
+def _update_1d_preconditioner(
+    q: torch.Tensor,
+    L: torch.Tensor,
+    precond_grad_cov: torch.Tensor,
+    total_numel: int,
+    precond_lr: float,
+    betaL: float,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Update 1D preconditioner with adaptive Lipschitz constant.
+
+    Args:
+        q: Kronecker factor 1D tensor for this dimension to update.
+        L: Lipschitz constant for this dimension.
+        precond_grad_cov: Gradient covariance.
+        total_numel: Total number of elements in the gradient.
+        precond_lr: Learning rate.
+        betaL: Beta for the Lipschitz constant exponential moving average.
+
+    Returns:
+        q: Updated Kronecker factor 1D tensor for this dimension.
+        L: Updated Lipschitz constant for this dimension.
+    """
+    normalization = total_numel / q.numel()
+    ell = torch.max(torch.real(precond_grad_cov)) + normalization
+    L = torch.max(betaL * L + (1 - betaL) * ell, ell)
+    q = q * (1 - precond_lr / L * (precond_grad_cov - normalization))
+    return q, L
+
+
+def _get_precond_lr(precond_lr: float, step: int, min_precond_lr: float = 0.3, warmup_steps: int = 10000) -> float:
+    r"""Helper function to get preconditioner learning rate for this optimization step based on a square root schedule.
+
+    Decaying from a higher lr down to min_precond_lr improves accuracy.
+
+    Args:
+        precond_lr: Learning rate.
+        step: Current step.
+        min_precond_lr: Minimum learning rate.
+        warmup_steps: Warmup steps.
+
+    Returns:
+        The preconditioner learning rate.
+    """
+
+    scheduled_lr = precond_lr / math.sqrt(1.0 + step / warmup_steps)
+    return max(scheduled_lr, min_precond_lr)
+
+
+def _dampen_tensor(x: torch.Tensor, scale: float) -> torch.Tensor:
```

Comment:
Suggestion: if this function is only called at one place, it is simple enough to be inline than a function.